### PR TITLE
subsys: console: Factor out fifo-based console input abstraction

### PR DESCRIPTION
**Pull request by pfalcon**
Console subsystem is intended to be a layer between console drivers
and console clients, like e.g. shell. This change factors out code
from shell which dealed with individial console drivers and moves it
to console subsystem, under the name console_register_line_input().

To accommodate for this change, older console subsys Kconfig symbol
is changed from CONFIG_CONSOLE_PULL to CONFIG_CONSOLE_SUBSYS
(CONFIG_CONSOLE is already used by console drivers). This signifies
that console subsystem is intended to deal with all of console
aspects in Zephyr (existing and new), not just provide some "new"
functionality on top of raw console drivers, like it initially
started.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

**Comment by pfalcon** [comment]
@andrewboie : This is 2nd change based on out console subsys discussion, following my comment https://github.com/zephyrproject-rtos/zephyr/pull/6815#issuecomment-394303018 :

> And what is in drivers/consoles/ are just disparate drivers, look at this ugliness: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/shell/shell.c#L635

Note that ideally and eventually, I'd like to get rid of this fifo-based input handling altogether. But of course, it's neither possible nor makes sense to get rid of it in one jump, because it's important functionality used by the actual code. So, it would need to be rewritten by other means, and maybe long eventually deprecated and removed, when no users left.

But in the meantime, it's indeed somewhat overdue to evolve the console *subsystem* to abstract users of console from individual console drivers, so let's do that now.

File: subsys/console/Kconfig [hunk 1]
```diff
@@ -4,16 +4,16 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
-menu "Console (pull-style)"
+menu "Console"
 
-config CONSOLE_PULL
```
Comment (ulfalizer) [inline subsys/console/Kconfig]: Still got `CONFIG_CONSOLE_PULL` assignments here:

 - `samples/subsys/console/echo/prj.conf`
 - `samples/subsys/console/getchar/prj.conf`
 - `samples/subsys/console/getline/prj.conf`

**Review by ulfalizer — CHANGES_REQUESTED** [review]

**Comment by andrewboie** [comment]
Do you have a notion of what the full console subsystem API will look like?

I'm thinking about:
- How we can wire this to printk()
- How we can do buffering before we make system calls to the console device drivers (uart, RAM console, etc), we need to abstract away what we have right now calling directly into the UART driver
- What changes we need to make to console device drivers to support our needs, the current __printk_hook_install stuff has to go. I know that I will need an API to write a C string once the buffer is full, to replace the current mechanism where printk() does its own buffering and then makes a k_str_out() syscall

**Comment by pfalcon** [comment]
> Do you have a notion of what the full console subsystem API will look like?

Yes, in my view, console_getchar() and console_putchar() should be primary conceptual interface of it, everything else should be implemented in terms of it. It should be generalized though, e.g. as ca9a98e6e2589503051a3eb7d2192037435c9fd9 shows, there's no consistency of how "would block" condition is handled across get/put, apparently those calls should take standard Zephyr timeout param, including NO_WAIT and FOREVER.

You may say that user/kernel context switching per char is expensive, and it should work with strings. That can be done, but adds its complication, e.g. how timeouts are handled. We would be very close to POSIX c_cc[VMIN] / c_cc[VTIME], and then very close to /dev/tty at all.

> How we can wire this to printk()

It would just call console_putchar()

> How we can do buffering before we make system calls to the console device drivers (uart, RAM console, etc), we need to abstract away what we have right now calling directly into the UART driver

I'd separate buffering to make user/kernel switching more efficient (like printk does currently) from I/O buffering (like subsys/console/getchar does). 1st is up to you. To answer the question on I/O buffering fully, we'd need to set requirements for multiple console driver support/multiplexing. To give one idea, input buffering could happen in console subsystem, but any output buffering is up to a driver. That would be the easiest way to support multiple input and multiple output devices, if we need that.

> What changes we need to make to console device drivers to support our needs, the current __printk_hook_install stuff has to go.

They should be made to work with getchar/putchar model. Everything else should be moved where it belongs. For example, current drivers/console/uart_console has line editing code. It doesn't make any sense, that code belongs to the console subsystem, reused for all drivers. How exactly drivers should work with getchar/putchar model depends again on multiple sources/sinks requirements. If there's no support for multiplicity, then a driver can just implement getchar/putchar directly (and then input buffering also belongs to it). Otherwise, there apparently should be a console subsystem "callback" to let a driver feed a new char into getchar queue. But puchar again maps 1:1, but we'd need to think about flow control (e.g. if one of drivers can't accept putchar(), what do we do - wait on it, or drop on it).

File: subsys/console/Kconfig (hunk 1 above)
Comment (pfalcon) [reply subsys/console/Kconfig]: Fixed, thanks.

**Review by ulfalizer — APPROVED** [review]
Leftover `CONFIG_CONSOLE_PULL` assignments fixed.

**Comment by andrewboie** [comment]
> I'd separate buffering to make user/kernel switching more efficient (like printk does currently) from I/O buffering (like subsys/console/getchar does)

Good distinction.

> You may say that user/kernel context switching per char is expensive, and it should work with strings. That can be done, but adds its complication

Worth doing IMO. It really will add a lot of overhead to do a syscall per-character. A variant of console_getchar() that takes a C string and a size as a syscall would be worth doing (pretty much what k_str_out() does), internally it could simply loop over the string contents and call console_getchar() on each one, on the kernel side of the privilege boundary.

> But puchar again maps 1:1, but we'd need to think about flow control (e.g. if one of drivers can't accept putchar(), what do we do - wait on it, or drop on it).

I haven't given this much thought, but at least right now k_str_out() just iterates over the string and calls _char_out() on it, which would block until the driver can actually emit the string. Maybe for printk() at least could be enough, to block until the I/O buffering flushes enough to take new characters. There's always alternative consoles like RAM console for situations where we can't have any overhead.

Or in other words, make this a parameter or something, in the case of printk() wait instead of dropping.

**Comment by pfalcon** [comment]
@ulfalizer: I just added a commit here to fix the native_posix build, "Don't make CONSOLE_GETCHAR and *_GETLINE exclusive". Can you please eyeball it that it's ok from your side? Thanks.

**Comment by ulfalizer** [comment]
@pfalcon 
Another possibility is to make the choice `optional`. The whole choice can then be disabled, like for a symbol.

See https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt:

    A choice accepts another option "optional", which allows to set the
    choice to 'n' and no entry needs to be selected.
    If no [symbol] is associated with a choice, then you can not have multiple
    definitions of that choice. If a [symbol] is associated to the choice,
    then you may define the same choice (ie. with the same entries) in another
    place.

**Review by ulfalizer — CHANGES_REQUESTED** [review]
Just a dummy review to make sure the `optional` choice thingy gets seen before this gets merged.

**Comment by pfalcon** [comment]
@ulfalizer:

> Another possibility is to make the choice optional.

Ok, I updated it to use `optional` instead. Please re-review.

**Review by ulfalizer — APPROVED** [review]
LGTM

File: subsys/console/line_fifo.c [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * @file
+ * @brief Legacy fifo-based line input
+ */
+
+#include <zephyr.h>
+#include <console.h>
+
+#ifdef CONFIG_UART_CONSOLE
+#include <drivers/console/uart_console.h>
```
Comment (andrewboie) [inline subsys/console/line_fifo.c]: argh, this is really gross.
Other subsystems define an API struct which would have a function pointer for "register_input"
Ifdefing per console driver like this doesn't scale.
I guess we can fix this in another patch.

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by finikorg** [comment]
I am thinking can we use common API to register shell input instead of current method, something like this: #8490

**Comment by pfalcon** [comment]
> I am thinking can we use common API to register shell input instead of current method, something like this: #8490

Your PR is concerned with "shell input", my PR is concerned with "console input". Please see above - the problem we currently have is that console clients (e.g. shell) call directly into console drivers. Instead, all console clients should call into the console subsystem, and that's what this patch does (and only that, actual things the moved code does remains the same, "one change per PR"). If after this PR, you'll want to implement changes like done in #8490 - I personally don't have any issues with that.

File: subsys/console/line_fifo.c (hunk 2 above)
Comment (nashif) [reply subsys/console/line_fifo.c]: agree, the registration needs to happen from the console driver, not here, lets fix this in another PR

**Review by nashif — APPROVED** [review]

**Pull request merged**